File: src/lib/ipc/license.ts
import { invoke } from '@tauri-apps/api/core';

/** Decoded payload returned by the backend when a license verifies.
 *  Mirrors `crate::license::Payload` 1:1. */
export interface LicensePayload {
  user: string;
  issued: string;
  expires?: string | null;
  features: string[];
  /** Set when the seller minted the license bound to a specific
   *  install. `null` / undefined = portable. */
  machine_id?: string | null;
}

/** Tagged union the backend returns for any license query. The
 *  React side switches on `kind` to decide which UI to render. */
export type LicenseVerdict =
  | { kind: 'missing' }
  | { kind: 'invalid'; reason: string }
  | { kind: 'expired'; user: string; expires: string }
  | {
      kind: 'wrong_machine';
      user: string;
      /** Machine id baked into the license token. */
      expected: string;
      /** Local install's machine id — what we want the seller to
       *  re-issue against. */
      actual: string;
    }
  | { kind: 'valid'; payload: LicensePayload };

export interface LicenseStatusReply {
  verdict: LicenseVerdict;
  /** True for `cargo build` debug binaries — frontend uses this to
   *  bypass the gate during development without disabling the IPC. */
  dev_mode: boolean;
}

/** Read + verify the on-disk license. Cheap; called once on app
 *  boot and after Settings → "Sign out". */
export function licenseStatus(): Promise<LicenseStatusReply> {
  return invoke<LicenseStatusReply>('license_status');
}

/** Verify + persist a token the user pasted into the activation
 *  dialog. Returns the verdict from `license_status` semantics so
 *  the frontend can render the new state without a follow-up call. */
export function licenseInstall(token: string): Promise<LicenseStatusReply> {
  return invoke<LicenseStatusReply>('license_install', { token });
}

/** Wipe the on-disk license. Idempotent — non-existent file is OK. */
export function licenseClear(): Promise<void> {
  return invoke<void>('license_clear');
}

/** Persistent per-install UUID. Shown in the gate + Settings so the
 *  user can email it to the seller, who mints a license bound to it. */
export function licenseMachineId(): Promise<string> {
  return invoke<string>('license_machine_id');
}

File: src-tauri/src/ipc/license.rs
//! License-key IPC. Three commands the frontend uses:
//!   - `license_status` — current verdict, called on app boot to
//!     decide whether to show the gate.
//!   - `license_install` — verify + persist a token the user pasted.
//!   - `license_clear` — wipe the on-disk file (Settings "Sign out").
//!
//! Each command is `spawn_blocking` because the verification touches
//! the filesystem; `verify_token` itself is fast (a single ed25519
//! check) but we don't want to block the Tauri runtime thread on a
//! slow network drive.

use serde::Serialize;
use tauri::State;

use crate::error::{IpcError, IpcResult};
use crate::license::{self, Verdict};
use crate::state::AppState;

/// Wrapper so the frontend gets `{ verdict, dev_mode }` and can
/// short-circuit the gate on dev builds without a separate IPC.
#[derive(Debug, Serialize)]
pub struct LicenseStatusReply {
    pub verdict: Verdict,
    /// True when the binary was built in debug mode. Frontend uses
    /// this to render a "DEV BUILD — license bypassed" banner instead
    /// of the full-screen gate. Production builds always send
    /// `dev_mode: false` so the gate is enforced.
    pub dev_mode: bool,
}

#[tauri::command]
pub async fn license_status(state: State<'_, AppState>) -> IpcResult<LicenseStatusReply> {
    let dir = state.config_dir.clone();
    let verdict = tokio::task::spawn_blocking(move || license::status(&dir))
        .await
        .map_err(|e| IpcError::Internal {
            message: format!("license_status join: {e}"),
        })?;
    Ok(LicenseStatusReply {
        verdict,
        dev_mode: cfg!(debug_assertions),
    })
}

#[tauri::command]
pub async fn license_install(
    state: State<'_, AppState>,
    token: String,
) -> IpcResult<LicenseStatusReply> {
    let dir = state.config_dir.clone();
    let token = token.trim().to_string();
    if token.is_empty() {
        return Err(IpcError::NotConfigured {
            hint: "license token is empty".into(),
        });
    }
    let verdict = tokio::task::spawn_blocking(move || license::install(&dir, &token))
        .await
        .map_err(|e| IpcError::Internal {
            message: format!("license_install join: {e}"),
        })?
        .map_err(|e| IpcError::Internal {
            message: format!("license_install: {e}"),
        })?;
    Ok(LicenseStatusReply {
        verdict,
        dev_mode: cfg!(debug_assertions),
    })
}

/// Return this install's persistent machine id. The activation UI
/// shows it BEFORE a license is installed so the user can email /
/// IM it to the seller, who mints a license bound to that exact id.
#[tauri::command]
pub async fn license_machine_id(state: State<'_, AppState>) -> IpcResult<String> {
    let dir = state.config_dir.clone();
    tokio::task::spawn_blocking(move || license::machine_id(&dir))
        .await
        .map_err(|e| IpcError::Internal {
            message: format!("license_machine_id join: {e}"),
        })
}

#[tauri::command]
pub async fn license_clear(state: State<'_, AppState>) -> IpcResult<()> {
    let dir = state.config_dir.clone();
    tokio::task::spawn_blocking(move || license::clear(&dir))
        .await
        .map_err(|e| IpcError::Internal {
            message: format!("license_clear join: {e}"),
        })?
        .map_err(|e| IpcError::Internal {
            message: format!("license_clear: {e}"),
        })?;
    Ok(())
}
